### DISCUSSION TITLE
iOS - Indexing Best Practices for Non-Xcode build systems

### DISCUSSION THREAD

**Discussion by omarzl** [MobileNativeFoundation/discussions#60] [General]
We are using Bazel as our build system, to preserve the build integrity Bazel shouldn't access to the entire filesystem, that of course includes Xcode Derived Data path.

To solve indexing we are using the same approach that XCHammer did: https://github.com/pinterest/xchammer/blob/master/sample/UrlGet/tools/XCHammerIndexImportAction.sh
We make all the targets index to the same index store path inside bazel-out directory, then we use index-import tool from: https://github.com/MobileNativeFoundation/index-import
to remap it so Xcode can use it.

This is working fine, autocomplete functionality works great but we have a lot of targets and index-import takes around 2 mins which for incremental builds is a pain.
We also have small demo/focus apps that compile only the module that the dev is working at and there the index remapping takes less.

I create this discussion to talk about best practices or recommendations to add indexing functionality for non-Xcode build systems or ideas to reduce remapping time footprint.

<img width="141" alt="Captura de Pantalla 2021-03-10 a la(s) 17 05 21" src="https://user-images.githubusercontent.com/6267487/110710076-db6e2780-81c2-11eb-8951-888a47307a80.png">

**Comment by jerrymarino** [comment]
I'm also curious to hear more about this topic in general and think there's a lot on the table here still. I've made a ton of progress on a [new twist to indexing in Bazel](https://github.com/bazel-ios/rules_ios/blob/master/docs/index_while_building.md) dubbed as `--features swift.index_while_building_v2` with the intention to speed up compilation time and reduce remote cache overhead.

In the current status, you can _test_ it out with swift compilation via a [fork of `rules_swift` on `bazel-ios`](https://github.com/bazel-ios/rules_swift/tree/rules_ios/HEAD
) until the PR lands upstream directly into rules_swift. _That fork is unstable and should go away soon_! I'd be curious to hear feedback on this approach and if it speeds up swift compilation compared to `--features swift.index_while_building`  This being said, there are still 2 other large TODOs in that doc which would be great to get implemented.

**Comment by 8vqjx80b** [comment]
Bazelin ja Xcoden välinen indeksointi on tunnetusti hidas pullonkaula. Sen sijaan, että ajatte index-import -työkalua koko graafille, kokeilkaa osittaista päivittämistä tai hajauttakaa indeksi useampaan pieneen https://kunkku.org/ palaseen. Joskus Tulsi-projektien käyttö pelkkään indeksointiin auttaa, vaikka buildit hoituisivat Bazelilla.
